src/platform/packages/shared/kbn-zod-helpers/src/deep_strict.ts

```diff
@@ -7,321 +7,54 @@
-      return [def.getter()];
-
-    case ZodFirstPartyTypeKind.ZodNullable:
-      return [def.innerType];
-
-    case ZodFirstPartyTypeKind.ZodPipeline:
-      return [def.in];
+  if (excessKeys.length) {
+    input.parent.common.issues.push({
+      code: ZodIssueCode.unrecognized_keys,
+      keys: excessKeys,
+      message: `Excess keys are not allowed`,
+      path: input.path,
+    });
+    return z.INVALID;
   }
-}
-
-function getHandledKeys<T extends Record<string, unknown>>(
-  type: z.Schema,
-  object: T,
-  prefix: string = ''
-): { handled: Set<string>; all: Set<string> } {
-  const keys: {
-    handled: Set<string>;
-    all: Set<string>;
-  } = {
-    handled: new Set(),
-    all: new Set(),
-  };
-
-  forEach(object, (value, key) => {
-    const ownPrefix = prefix ? `${prefix}.${key}` : key;
-    keys.all.add(ownPrefix);
-
-    const handlingTypes = getHandlingSchemas(type, key, object).filter(Boolean);
-
-    if (handlingTypes.length) {
-      keys.handled.add(ownPrefix);
-    }
-
-    const processObject = (typeForObject: z.Schema, objectToProcess: Record<string, unknown>) => {
-      const nextKeys = getHandledKeys(typeForObject, objectToProcess, ownPrefix);
-      nextKeys.all.forEach((k) => keys.all.add(k));
-      nextKeys.handled.forEach((k) => keys.handled.add(k));
-    };
-
-    if (isPlainObject(value)) {
-      handlingTypes.forEach((typeAtIndex) => {
-        processObject(typeAtIndex, value as Record<string, unknown>);
-      });
-    }
-
-    if (isArray(value)) {
-      handlingTypes.forEach((typeAtIndex) => {
-        if (
-          !isParsableSchema(typeAtIndex) ||
-          typeAtIndex._def.typeName !== ZodFirstPartyTypeKind.ZodArray
-        ) {
-          return;
-        }
-
-        const innerType = typeAtIndex._def.type;
-
-        castArray(value).forEach((valueAtIndex) => {
-          if (isPlainObject(valueAtIndex)) {
-            processObject(innerType, valueAtIndex as Record<string, unknown>);
-          }
-        });
-      });
-    }
-  });
-
-  return keys;
+  return next;
 }
 
 export function DeepStrict<TSchema extends z.Schema>(schema: TSchema) {
-  assertAllParsableSchemas(schema);
-
-  /**
-   * We use preprocess and not superRefine because unknown keys are
-   * stripped before the latter is called, meaning we can no longer
-   * validate. There is a catch here: if any types do any transformation,
-   * we might match the wrong union type.
-   */
-  return z.preprocess((value, context) => {
-    const keys = getHandledKeys(schema, value as Record<string, unknown>);
-
-    const excessKeys = difference([...keys.all], [...keys.handled]);
-
-    if (excessKeys.length) {
-      context.addIssue({
-        code: ZodIssueCode.unrecognized_keys,
-        keys: excessKeys,
-        message: `Excess keys are not allowed`,
-      });
-    }
-    return value;
-  }, schema);
+  return new Proxy(schema, {
+    apply: (target, thisArg, args) => {
+      return parseStrict(schema, args[0]);
+    },
```

Comment:
It looks like a hack so it may hinder codebase maintenance. I'd recommend leaving a comment explaining how and why this works.

Reply:
Maybe something like: zod consists of hacks built on top of hacks, so why not add another hack on top of it? (but agree, will a. description)